### PR TITLE
inference: allocate CodeInstance with inference results ASAP

### PR DESCRIPTION
**Pull request by vtjnash**
Update the CodeInstance after optimizations with the code (if possible), but construct it early so that there will be the option of using it for edges and invoke during optimizations, without problems with cycles.

@Keno the intent is that this will let us use CodeInstance as targets for `:invoke` calls, as discussed in https://hackmd.io/bpLnKm48TbWP_U_PGWgEVg. It complicates the life-cycle of a CodeInstance a little bit, so I don't think we have yet fully specified the thread-safety of this and which are valid transitions here (in general, it is only legal to narrow these properties or widen the worlds), but I think that is okay to postpone.

### CODE REVIEW

File: base/compiler/compiler.jl [hunk 1]
```diff
@@ -14,6 +14,7 @@ const setproperty! = Core.setfield!
 const swapproperty! = Core.swapfield!
 const modifyproperty! = Core.modifyfield!
 const replaceproperty! = Core.replacefield!
+const _DOCS_ALIASING_WARNING = ""
```
Comment (aviatesk) [inline base/compiler/compiler.jl:17]:
```suggestion
```
Looks like an unrelated change?

Comment (vtjnash) [reply base/compiler/compiler.jl:17]: True. We should land it eventually, as it is required for loading a testing version of inference now, but it isn't connected to any particular PR.

File: src/julia.h [hunk 2]
```diff
@@ -457,15 +457,6 @@ typedef struct _jl_code_instance_t {
 
     // purity results
     // see also encode_effects() and decode_effects() in `base/compiler/effects.jl`,
-    uint32_t ipo_purity_bits;
```
Comment (Keno) [inline src/julia.h:460]: keep the `ipo_purity_bits` name, please. We're not doing any non-IPO optimizations at the moment, but we may in the future, so there's no reason to rename it now.

Comment (vtjnash) [reply src/julia.h:460]: I was keeping the shorter name, due to discussion with @aviatesk. Non-IPO optimizations would presumably be done locally anyways, and not require this global metadata?

Comment (Keno) [reply src/julia.h:460]: No, we may need to cache them if we want to allow non-IPO inlining which will be desirable for our float functions once we properly effect model them rather than pretending they're consistent.

Comment (vtjnash) [reply src/julia.h:460]: What is non-interprocedural inlining and why doesn't it require source code?

Comment (Keno) [reply src/julia.h:460]: I meant that as a shorthand. In general, you may want to re-run the full optimization pipeline in non-IPO mode, including concrete eval (which needs to know the non-IPO variant of the bits) DCO (which also needs it) and inlining (which needs a non-IPO version of the source as you point out). Perhaps inlining is confusing because it only really make sense if the non-IPO version is inlineable, but the IPO version is not.

Comment (vtjnash) [reply src/julia.h:460]: So are you just saying you want the longer name kept for now instead? It seemed like all the names here refer to the IPO info, so it seemed redundant, but I can switch which name is kept if that is all you meant

Comment (Keno) [reply src/julia.h:460]: Yes, keep the name

File: src/jltypes.c [hunk 3]
```diff
@@ -3506,11 +3506,11 @@ void jl_init_types(void) JL_GC_DISABLED
...
-    const static uint32_t code_instance_atomicfields[1] = { 0b110110101100011100 }; // Set fields 3-5, 9, 10, 12, 14-15, 17-18 as atomic
-    //Fields 4-5 are only operated on by construction and deserialization, so are const at runtime
-    //Fields 13 and 17 must be protected by locks, and thus all operations on jl_code_instance_t are threadsafe
-    //Except for field 9 (inferred), which is volatile unless you know which other places are currently using it
+    const static uint32_t code_instance_constfields[1]  = { 0b00000100011100011 }; // Set fields 1, 2, 6-8, 12 as const
+    const static uint32_t code_instance_atomicfields[1] = { 0b11011011100011100 }; // Set fields 3-5, 9, 10, 11, 13-14, 16-17 as atomic
+    // Fields 4-5 are only operated on by construction and deserialization, so are effectively const at runtime
+    // Fields ipo_purity_bits and analysis_results are not currently threadsafe or reliable, as they get mutated after optimization, but are not declared atomic
+    // and there is no way to tell (during inference) if their value is finalized yet (to wait for them to be narrowed if applicable)
```
Comment (aviatesk) [inline src/jltypes.c:3513]:
```suggestion
    const static uint32_t code_instance_atomicfields[1] = { 0b11011001100011100 }; // Set fields 3-5, 9, 10, 11, 13-14, 16-17 as atomic
    // Fields 4-5 are only operated on by construction and deserialization, so are effectively const at runtime
    // Fields ipo_purity_bits and analysis_results are not currently threadsafe or reliable, as they get mutated after optimization, but are not declared atomic
    // and there is no way to tell (during inference) if their value is finalized yet (to wait for them to be narrowed if applicable)
```
According to the comment on L3512, it seems that `ipo_purity_bits` should not be marked as atomic at this point, doesn't it?

File: base/compiler/typeinfer.jl [hunk 4]
```diff
@@ -442,15 +447,18 @@ function cache_result!(interp::AbstractInterpreter, result::InferenceResult)
...
     if typ isa LimitedAccuracy
         if sv.parent === nothing
-            # when part of a cycle, we might have unintentionally introduced a limit marker
-            @assert !isempty(sv.callers_in_cycle)
+            # we might have introduced a limit marker, but we should know it must be sv and other callers_in_cycle
+            #@assert !isempty(sv.callers_in_cycle)
+            #  FIXME: this assert fails, appearing to indicate there is a bug in filtering this list earlier.
+            #  In particular (during doctests for example), during inference of
+            #  show(Base.IOContext{Base.GenericIOBuffer{Memory{UInt8}}}, Base.Multimedia.MIME{:var"text/plain"}, LinearAlgebra.BunchKaufman{Float64, Array{Float64, 2}, Array{Int64, 1}})
+            #  we observed one of the ssavaluetypes here to be Core.Compiler.LimitedAccuracy(typ=Any, causes=Core.Compiler.IdSet(getproperty(LinearAlgebra.BunchKaufman{Float64, Array{Float64, 2}, Array{Int64, 1}}, Symbol)))
```
Comment (aviatesk) [inline base/compiler/typeinfer.jl:461]: AFAIU, this assertion should be failing because of the removal of `push!(frames, frame)` on L239, i.e. it fails when `LimitedAccuracy` is introduced in cases of self-recursion. Was this an intentional change?

File: base/compiler/typeinfer.jl [hunk 5]
```diff
@@ -250,15 +287,13 @@ end
...
+    finishinfer!(frame, frame.interp)
     opt = frame.result.src
     if opt isa OptimizationState # implies `may_optimize(caller.interp) === true`
         optimize(frame.interp, opt, frame.result)
     end
     finish!(frame.interp, frame)
-    if is_cached(frame)
-        cache_result!(frame.interp, frame.result)
-    end
+    unlock_mi_inference(frame.interp, frame.linfo)
```
Comment (aviatesk) [inline base/compiler/typeinfer.jl:296]: Maybe `unlock_mi_inference` can be incorporated into `finish!`?

File: base/compiler/typeinfer.jl (hunk 4 above)
Comment (vtjnash) [reply base/compiler/typeinfer.jl:461]: The code at L239 seemed to have been making this accidentally pass and hiding this bug

File: src/jltypes.c (hunk 3 above)
Comment (vtjnash) [reply src/jltypes.c:3513]: That is possibly true, but we don't have a correct mechanism right now to avoid that data-race except marking it atomic.

Comment (aviatesk) [reply src/jltypes.c:3513]: So it sounds probably better to mark `analysis_results` atomic too?

Comment (vtjnash) [reply src/jltypes.c:3513]: Probably, though I commonly do not fix the atomic marking of the fields in the file unless I am specifically changing them for some other purpose
